### PR TITLE
`DataExporter : Sync`

### PR DESCRIPTION
**Pull request by robertbastian**

### CODE REVIEW

**Comment by sffc** [comment]
How does this relate to #1600?

File: provider/core/src/hello_world.rs
Comment (sffc) [inline provider/core/src/hello_world.rs:188]: Issue: I do not agree that this impl is "weird".  It is a useful example of how data can be exported into a memory store as opposed to serialized into a blob or filesystem.  Please restore the impl.

File: provider/blob/src/export/blob_exporter.rs
Comment (sffc) [inline provider/blob/src/export/blob_exporter.rs:23]: Issue: If people are forced to write `unsafe impl` when implementing DataExporter, it means there's a problem with DataExporter.  `unsafe impl` can be very unsafe and is hard to debug.  Since the `sink` field is only used in the `close()` method, maybe we need to pass more context into that method.  Or perhaps the `sink` field needs to be changed in a way that makes it threadsafe; can you make it wrap a `dyn Write + Sync` maybe?

File: provider/fs/src/export/aliasing.rs
Comment (sffc) [inline provider/fs/src/export/aliasing.rs:43]: Optional: If it makes things easier, feel free to delete this file and the symlink aliasing features.  This was always fraught.

File: provider/fs/src/export/fs_exporter.rs
Comment (sffc) [inline provider/fs/src/export/fs_exporter.rs:62]: Suggestion: This looks far more complicated than it's worth. I think we should remove this from FsDataProvider and have files always written out. Deduplication belonds in BlobProvider.

File: provider/fs/src/export/mod.rs
Comment (sffc) [inline provider/fs/src/export/mod.rs:60]: Issue: why is testdata.postcard changed? Try to make this a no-op change on testdata.postcard.

**Review by sffc — CHANGES_REQUESTED** [review]

File: provider/blob/src/export/blob_exporter.rs
Comment (Manishearth) [reply provider/blob/src/export/blob_exporter.rs:23]: Big +1 here, we should really not be using unsafe impls here. Let's wrap the sink field.

You basically should never be doing `unsafe impl` for `Sync` or `Send` unless you're implementing something low level or doing some novel strategy. We're just using mutexes and such here, it should just work.

Comment (Manishearth) [reply provider/blob/src/export/blob_exporter.rs:23]: Alternatively, we can have the `&mut self` method that writes to the sink take the sink as a parameter so that the exporter does not hold on to the actual sink.

Comment (robertbastian) [reply provider/blob/src/export/blob_exporter.rs:23]: The close method cannot accept the sink because it's a blob specific argument. I didn't want to add the `Sync` restriction on sink because it's not needed. I've since discovered `Mutex::get_mut` which doesn't lock because it has a mut ref, so I'll just do that.

File: provider/fs/src/export/mod.rs
Comment (robertbastian) [reply provider/fs/src/export/mod.rs:60]: Because there's a key collision and we're now adding the data to the `ZeroMap2d` in a different order :). See #1622

File: provider/core/src/hello_world.rs
Comment (robertbastian) [reply provider/core/src/hello_world.rs:188]: It's not compatible with the changes I'm introducing here though. To make `DataExporter` `Sync` it needs to use internal mutability, as `put_payload` becomes non-mut. Do you want me to introduce a `RwLock` in `HelloWorldProvider` just so that this continues working?

Comment (robertbastian) [reply provider/core/src/hello_world.rs:188]: Actually we don't even have locks in `no_std`

**Review by sffc — APPROVED** [review]
OK.  Please link an issue.  I prefer if you wait for @Manishearth's approval, too.

**Review by Manishearth — APPROVED** [review]

**Pull request merged**
